cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2209,6 +2209,45 @@ TEST(Cast, ListToListOptionsPassthru) {
   }
 }
 
+static void CheckFSLToFSL(const std::vector<std::shared_ptr<DataType>>& value_types,
+                          const std::string& json_data) {
+  for (const auto& src_value_type : value_types) {
+    for (const auto& dest_value_type : value_types) {
+      const auto src_type = fixed_size_list(src_value_type, 2);
+      const auto dest_type = fixed_size_list(dest_value_type, 2);
+      ARROW_SCOPED_TRACE("src_type = ", src_type->ToString(),
+                         ", dest_type = ", dest_type->ToString());
+      auto src_array = ArrayFromJSON(src_type, json_data);
+      CheckCast(src_array, ArrayFromJSON(dest_type, json_data));
```

Comment:
I guess we should also test casting slices of the input and comparing it to slices of the output here.

Reply:
We should also slice the child array, construct the parent array (with nulls, etc.) and test that as well. That should probably be a separate test/helper function. There's a helper function called TweakValidityBit that is useful for these sorts of things

Reply:
I think CheckCast covers these interesting cases already (which ultimately calls the following function).

https://github.com/apache/arrow/blob/6cc37cf2d1ba72c46b64fbc7ac499bd0d7296d20/cpp/src/arrow/compute/kernels/test_util.cc#L106-L172

Reply:
Have also added test with `TweakValidityBit`.

Reply:
It doesn't cover a case like this:

```cpp
auto children = ArrayFromJSON(int16(), "[1, 2, null, 4, 5, null]");
children = children->Slice(2);
// Omitting null bitmap here, but we should test that as well
auto fsl = std::make_shared<FixedSizeListArray>(fixed_size_list(int16(), 2), 2, children);
// fsl is the list [[null, 4], [5, null]]
```

e.g. in Python:

```python
>>> pyarrow.FixedSizeListArray.from_arrays(pyarrow.array([1, 2, None, 4, 5, None]).slice(2), type=pyarrow.list_(pyarrow.int64(), 2))
<pyarrow.lib.FixedSizeListArray object at 0x7f44bc40f0a0>
[
  [
    null,
    4
  ],
  [
    5,
    null
  ]
]
```

Reply:
Thanks for the reference, @lidavidm. Will update the tests soon :)

Reply:
Have updated the test and hopefully covered the interesting case.

PTAL :(